File: solutions/haskell/perfect-numbers/3/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify num =
    if num < 1
    then Nothing
    else classify' num

classify' :: Int -> Maybe Classification
classify' num
    | factorsSum num < num = Just Deficient
    | factorsSum num > num = Just Abundant
    | otherwise = Just Perfect

factorsSum :: Int -> Int
factorsSum number = sum (factors number)

factors :: Int -> [Int]
factors number =
    [divisor | divisor <- [1 .. (number `quot` 2)], number `rem` divisor == 0]

File: solutions/clojure/perfect-numbers/3/src/perfect_numbers.clj
(ns perfect-numbers)

(defn factors-of
  [num]
  (->>
   (range 1 (inc (/ num 2)))
   (filter #(zero? (rem num %)))))

(defn classify
  [num]
  (if (neg? num)
    (throw (IllegalArgumentException. "Number shouldn't be negative."))
    (let [as (apply + (factors-of num))]
      (cond
        (< as num) :deficient
        (> as num) :abundant
        (= as num) :perfect))))
